internal/elasticsearch/index/data_stream_lifecycle.go

```diff
@@ -0,0 +1,232 @@
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func ResourceDataStreamLifecycle() *schema.Resource {
+	dataStreamLifecycleSchema := map[string]*schema.Schema{
+		"id": {
+			Description: "Internal identifier of the resource",
+			Type:        schema.TypeString,
+			Computed:    true,
+		},
+		"name": {
+			Description: "Name of the data stream. Supports wildcards (*). To target all data streams use * or _all.",
+			Type:        schema.TypeString,
+			Required:    true,
+		},
+		"expand_wildcards": {
+			Description: `Type of data stream that wildcard patterns can match. Supports comma-separated values, such as open,hidden. Valid values are:
+
+			all, hidden - Match any data stream, including hidden ones.
+			open, closed - Matches any non-hidden data stream. Data streams cannot be closed.
+			none - Wildcard patterns are not accepted.
+			 
+			Defaults to open.`,
+			Type:             schema.TypeString,
+			Default:          "open",
+			Optional:         true,
+			ValidateDiagFunc: utils.AllowedExpandWildcards,
+		},
+		"data_retention": {
+			Description: "If defined, every document added to this data stream will be stored at least for this time frame. Any time after this duration the document could be deleted. When empty, every document in this data stream will be stored indefinitely",
+			Type:        schema.TypeString,
+			Optional:    true,
+		},
+		"enabled": {
+			Description: "If defined, it turns data stream lifecycle on/off (true/false) for this data stream. A data stream lifecycle that is disabled (enabled: false) will have no effect on the data stream. Defaults to true.",
+			Type:        schema.TypeBool,
+			Optional:    true,
+			Default:     true,
+		},
+		"downsampling": {
+			Description: "An optional array of downsampling configuration objects, each defining an after interval representing when the backing index is meant to be downsampled (the time frame is calculated since the index was rolled over, i.e. generation time) and a fixed_interval representing the downsampling interval (the minimum fixed_interval value is 5m). A maximum number of 10 downsampling rounds can be configured",
+			Type:        schema.TypeList,
+			Optional:    true,
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"after": {
+						Description: "Interval representing when the backing index is meant to be downsampled",
+						Type:        schema.TypeString,
+						Required:    true,
+					},
+					"fixed_interval": {
+						Description: "The interval at which to aggregate the original time series index. For example, 60m produces a document for each 60 minute (hourly) interval. This follows standard time formatting syntax as used elsewhere in Elasticsearch.",
+						Type:        schema.TypeString,
+						Required:    true,
+					},
+				},
+			},
+		},
+		"lifecycles": {
```

Comment:
It's not clear what value we get with this block. 

Since we're not writing these values back to the top level attributes TF drift detection won't pickup out of band changes. Choosing which value to use when the response includes multiple lifecycles is a challenge, but something we should solve IMO (a simple heuristic would be to use any value which differed from the configured value so drift detection worked, and if no diff was found just use the first). 

In reality, I suspect most users will likely target a single data stream.

Reply:
Originally I had it so that I only detected the value that changed, then I thought it would be nice to have everything in state, but I didn't know that TF couldn't work with it that way.

I will change it back :)